tests/python/pants_test/backend/python/rules/test_resolve_requirements.py

```diff
@@ -0,0 +1,68 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import json
+import zipfile
+
+from pants.backend.python.rules.resolve_requirements import (ResolvedRequirementsPex,
+                                                             ResolveRequirementsRequest,
+                                                             resolve_requirements)
+from pants.backend.python.subsystems.python_native_code import PexBuildEnvironment
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.engine.fs import Snapshot
+from pants.engine.rules import RootRule
+from pants.util.collections import assert_single_element
+from pants_test.test_base import TestBase
+
+
+class TestResolveRequirements(TestBase):
+
+  @classmethod
+  def rules(cls):
+    return super(TestResolveRequirements, cls).rules() + [
+      resolve_requirements,
+      RootRule(PythonSetup),
+      RootRule(PexBuildEnvironment),
+      RootRule(ResolveRequirementsRequest),
+    ]
+
+  def create_pex_and_get_pex_info(
+    self, requirements=None, entry_point=None, interpreter_constraints=None
+  ):
+    def hashify_optional_collection(iterable):
+      return tuple(sorted(iterable)) if iterable is not None else tuple()
+
+    request = ResolveRequirementsRequest(
+      requirements=hashify_optional_collection(requirements),
+      output_filename="test.pex",
+      entry_point=entry_point,
+      interpreter_constraints=hashify_optional_collection(interpreter_constraints),
+    )
+    requirements_pex = assert_single_element(
+      self.scheduler.product_request(ResolvedRequirementsPex, [request])
```

Comment:
Tests are coming together!

Got it down to this error now: `Exception: No installed @rules can compute ResolvedRequirementsPex for input Params(ResolveRequirementsRequest)`

I'm surprised by this because of the above lines 24 - 30 where we register `RootRule`s and the rule `resolve_requirements`. Any ideas?

Reply:
Reviewing from a phone, but the rule also requires a PexBuildEnvironment. Is that supposed to be constructable as a subsystem?

If so, take a look at `--native-engine-visualize-to` to see what this rule needs ...it probably needs either `OptionsBootstrapper` in order to construct those subsystems, or the subsystems themselves added as `Params`: you're set up to do the latter since you've added those as RootRules.

Reply:
This comment + @cosmicexplorer's help got me a bit closer to the answer! No longer complaining about `no installed @rules`.

Now, I get this error:

```
                    E       pants.engine.scheduler.ExecutionError: 1 Exception encountered:
                     E       Computing Select((<pants.backend.python.subsystems.python_native_code.PythonNativeCode object at 0x106ac6eb8>+<pants.backend.python.subsystems.python_setup.PythonSetup object at 0x106ac6e48>+ResolveRequirementsRequest(requirements<HashableTypedCollection(Exactly(str))>=(), output_filename<Exactly(str)>=test.pex, entry_point<Exactly(str or NoneType)>=None, interpreter_constraints<HashableTypedCollection(Exactly(str))>=('CPython>=2.7,<3', 'CPython>=3.6,<4'))), ResolvedRequirementsPex)
                     E         Computing Task(resolve_requirements(), (<pants.backend.python.subsystems.python_native_code.PythonNativeCode object at 0x106ac6eb8>+<pants.backend.python.subsystems.python_setup.PythonSetup object at 0x106ac6e48>+ResolveRequirementsRequest(requirements<HashableTypedCollection(Exactly(str))>=(), output_filename<Exactly(str)>=test.pex, entry_point<Exactly(str or NoneType)>=None, interpreter_constraints<HashableTypedCollection(Exactly(str))>=('CPython>=2.7,<3', 'CPython>=3.6,<4'))), ResolvedRequirementsPex, true)
                     E           Computing Task(create_pex_native_build_environment(), <pants.backend.python.subsystems.python_native_code.PythonNativeCode object at 0x106ac6eb8>, PexBuildEnvironment, true)
                     E             Throw(cpp_flags)
                     E               Traceback (most recent call last):
                     E                 File "/Users/eric/DocsLocal/code/projects/pants/.pants.d/pyprep/sources/256228084d76246646151fdd6fa86c9355906dce/pants_test/option/util/fakes.py", line 35, in __getattr__
                     E                   value = self._option_values[key]
                     E               KeyError: 'cpp_flags'
                     E       
                     E               During handling of the above exception, another exception occurred:
                     E       
                     E               Traceback (most recent call last):
                     E                 File "/Users/eric/DocsLocal/code/projects/pants/.pants.d/pyprep/sources/256228084d76246646151fdd6fa86c9355906dce/pants/engine/native.py", line 276, in call
                     E                   val = func(*args)
                     E                 File "/Users/eric/DocsLocal/code/projects/pants/.pants.d/pyprep/sources/256228084d76246646151fdd6fa86c9355906dce/pants/backend/python/subsystems/python_native_code.py", line 167, in create_pex_native_build_environment
                     E                   cpp_flags=[text_type(v) for v in python_native_code.get_options().cpp_flags],
                     E                 File "/Users/eric/DocsLocal/code/projects/pants/.pants.d/pyprep/sources/256228084d76246646151fdd6fa86c9355906dce/pants_test/option/util/fakes.py", line 38, in __getattr__
                     E                   raise AttributeError(key)
                     E               AttributeError: cpp_flags
```

I think maybe the `PythonNativeCode` subystem isn't being constructed properly, so its `cpp_flags` value does not exist? The stack trace mentions `fakes.py`.

Reply:
Ah... yea. There are some situations where the fake options used in tests require that you explicitly configure all options values that you will read. I've never been 100% certain which cases those are... but this looks like one. cc @benjyw

Reply:
I don't think there should be such cases if you create the test options via, e.g., `pants_test.subsystem.subsystem_util.init_subsystems`.

Reply:
`init_subsystems()` worked! Thank you Benjy!